qa/tasks/mon_netsplit.py

```diff
@@ -0,0 +1,205 @@
+"""
+Monitor netsplit testing
+"""
+import logging
+import contextlib
+import ceph_manager
+import time
+import gevent
+import itertools
+import re
+from teuthology import misc as teuthology
+from teuthology.orchestra import run
+
+log = logging.getLogger(__name__)
+
+from mon_thrash import _get_mons
+
+def get_mons(ctx):
+    return _get_mons(ctx)
+
+class MonitorNetsplit:
+    """
+    How it works::
+
+    - loop through the monitors
+    - for each:
+    -   kill its connection to next monitor
+    -   sleep for 'netsplit_period' seconds
+    -   restore connection
+    -   wait for quorum to be formed
+    -   sleep for 'netsplit_interval' seconds
+    -   repeat for the other monitors
+
+
+    Right now we just use iptables to black hole the IP address of each monitor from the other.
+    In future we might make it easier to blackhole monitors on shared hosts by specifying ports, but
+    that will require more work to cooperatively blackhole all the OSDs associated with them when you're
+    testing for datacenter splits.
+
+    Options::
+
+    netsplit_period     Number of seconds to blackhole the connection for (default: 10)
+    netsplit_interval   number of seconds to wait before blackholing the next connection (default: 10)
+
+    scrub               Scrub after each iteration (default: True)
+
+    For example::
+
+    tasks:
+    - ceph:
+    - mon_netsplit:
+        netsplit_period: 15
+        netsplit_interval: 5
+    - ceph-fuse:
+    - workunit:
+        clients:
+          all:
+            - mon/workloadgen.sh
+    """
+    def __init__(self, ctx, manager, config, logger):
+        self.ctx = ctx
+        self.manager = manager
+        self.manager.wait_for_clean()
+
+        self.stopping = False
+        self.logger = logger
+        self.config = config
+
+        if self.config is None:
+            self.config = dict()
+
+        """ Configs """
+        self.netsplit_period = float(self.config.get('netsplit_period', 10.0))
+        self.netsplit_interval = float(self.config.get('netsplit_interval', 10.0))
+        self.scrub = self.config.get('scrub', True)
+
+        self.thread = gevent.spawn(self.do_netsplit)
+
+    def log(self, x):
+        """
+        locally log info messages
+        """
+        self.logger.info(x)
+
+    def do_join(self):
+        """
+        Break out of this processes thrashing loop.
+        """
+        self.stopping = True
+        self.thread.get()
+
+    def do_netsplit(self):
```

Comment:
I suspect its more flexible to build this on top of a simpler tasks that *just* does the netsplit part, so you can do something like
```
tasks:
...
- netsplit.drop: [role1, role2]
...
- netsplit.undrop: [role1, role2]
```
The role might be mon.a or osd.2 or client.foo or whatever is in the roles section for the yaml.

The the mon_netsplit task would do the iteration over combinations... *or* you could write a simple test that specifically drops traffic between mon b and c and does some test in between.  We already have most of the pieces in the block below (like ceph.wait_for_healthy), wait for quorum, ability to run specific commands, etc.